### PR TITLE
feat(cts): add client test for retry strategy

### PR DESCRIPTION
**Pull request by millotp**
## 🧭 What and Why

[DI-1810](https://algolia.atlassian.net/browse/DI-1810)

Add a retry strategy test, which works by creating 2 servers, the first one should timeout, and the second one returns something we can check against, then we make sure we had the correct response, and that the timeout server was called.

This ensures 3 things in the client:
- The hosts are configurable
- The timeout is correct
- The retry strategy tries another host.

Note: this only works in js for now, but we can add it everywhere

[DI-1810]: https://algolia.atlassian.net/browse/DI-1810?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by morganleroi** [comment]
Whooo ! So nice, so fast !

File: clients/algoliasearch-client-scala/src/main/scala/algoliasearch/config/Host.scala [hunk 1]
```diff
@@ -8,5 +8,7 @@ package algoliasearch.config
   *   Whether this host should be used for read and/or write requests.
   * @param scheme
   *   Host protocol (i.e. `http`, `https`)
+  * @param port
+  *   Host port
   */
-case class Host(url: String, callTypes: Set[CallType], scheme: String = "https")
+case class Host(url: String, callTypes: Set[CallType], scheme: String = "https", port : Option[Int] = None)
```
Comment (Fluf22) [inline clients/algoliasearch-client-scala/src/main/scala/algoliasearch/config/Host.scala:14]: Sad the linter didn't catch it 😬 Not a mandatory change right now, though (can be updated in another PR)

```suggestion
case class Host(url: String, callTypes: Set[CallType], scheme: String = "https", port: Option[Int] = None)
```

**Review by Fluf22 — DISMISSED** [review]
Let's go! Great improvements in our testing strats

File: clients/algoliasearch-client-scala/src/main/scala/algoliasearch/config/Host.scala (hunk 1 above)
Comment (millotp) [reply clients/algoliasearch-client-scala/src/main/scala/algoliasearch/config/Host.scala:14]: the linter catches it in fact, the generated code will have the fix, it's just not pushed becaused of PHP 😠

Comment (Fluf22) [reply clients/algoliasearch-client-scala/src/main/scala/algoliasearch/config/Host.scala:14]: Perfect

File: clients/algoliasearch-client-java/algoliasearch/src/main/java/com/algolia/config/Host.java [hunk 2]
```diff
@@ -16,7 +17,14 @@ public Host(@Nonnull String url, @Nonnull Set<CallType> callType) {
   }
 
   public Host(String url, Set<CallType> callType, String scheme) {
-    this.url = url;
+    if (url.contains(":")) {
```
Comment (shortcuts) [inline clients/algoliasearch-client-java/algoliasearch/src/main/java/com/algolia/config/Host.java:20]: can this break if the user gave e.g. `http://...:1234` ?

File: generators/src/main/java/com/algolia/codegen/cts/tests/TestsClient.java [hunk 3]
```diff
@@ -63,10 +63,26 @@ public void run(Map<String, CodegenModel> models, Map<String, CodegenOperation>
...
+
+              stepOut.put("useEchoRequester", !hasCustomHosts);
+              stepOut.put("hasCustomHosts", hasCustomHosts);
+              if (hasCustomHosts && !language.equals("javascript") && !"true".equals(System.getenv("CI"))) {
+                // hack for docker on mac, the `network=host` does not work so we need to use
+                // another local IP
+                step.parameters.put(
+                  "customHosts",
+                  ((List<String>) step.parameters.get("customHosts")).stream()
+                    .map(host -> host.replace("localhost", "host.docker.internal"))
```
[the remark this answers is not in the document]
Comment (millotp) [reply generators/src/main/java/com/algolia/codegen/cts/tests/TestsClient.java:82]: #2525 could fix it

**Review by shortcuts — COMMENTED** [review]
that's a wonderful idea

File: clients/algoliasearch-client-java/algoliasearch/src/main/java/com/algolia/config/Host.java (hunk 2 above)
Comment (millotp) [reply clients/algoliasearch-client-java/algoliasearch/src/main/java/com/algolia/config/Host.java:20]: yes it will break, but there is a parameter for the `scheme` already

File: generators/src/main/java/com/algolia/codegen/cts/tests/TestsClient.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (shortcuts) [reply generators/src/main/java/com/algolia/codegen/cts/tests/TestsClient.java:82]: can we `if CI` in the meantime maybe?

[the remark this answers is not in the document]
Comment (millotp) [reply generators/src/main/java/com/algolia/codegen/cts/tests/TestsClient.java:82]: bah it's already a `if CI`, but the generated code will pushed with the change for the CI

File: clients/algoliasearch-client-java/algoliasearch/src/main/java/com/algolia/config/Host.java (hunk 2 above)
Comment (shortcuts) [reply clients/algoliasearch-client-java/algoliasearch/src/main/java/com/algolia/config/Host.java:20]: yup but I wonder if people wouldn't still do the mistake but I guess that's fine

File: clients/algoliasearch-client-php/lib/RetryStrategy/ApiWrapper.php [hunk 4]
```diff
@@ -148,7 +154,16 @@ private function request(
 
         $retry = 1;
         foreach ($hosts as $host) {
-            $uri = $uri->withHost($host);
+            if ($this->config->getHasFullHosts()) {
```
Comment (shortcuts) [inline clients/algoliasearch-client-php/lib/RetryStrategy/ApiWrapper.php:157]: so IIUC the user has to explicitly call `setFullHost` to make that work right? if that's the case we should provide some inlined doc to at least explain what we expect

File: clients/algoliasearch-client-python/algoliasearch/http/hosts.py [hunk 5]
```diff
@@ -6,9 +6,14 @@ class Host:
     TTL = 300.0
 
     def __init__(
-        self, url: str, priority: Optional[int] = 0, accept: Optional[int] = None
+        self,
+        url: str,
+        scheme="https",
```
Comment (shortcuts) [inline clients/algoliasearch-client-python/algoliasearch/http/hosts.py:11]:
```suggestion
        scheme: str ="https",
```

not sure if type inference works here

File: generators/src/main/java/com/algolia/codegen/cts/tests/TestsClient.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (shortcuts) [reply generators/src/main/java/com/algolia/codegen/cts/tests/TestsClient.java:82]: ahhh yes I missed that oki

File: templates/python/config.mustache [hunk 6]
```diff
@@ -51,12 +51,13 @@ class {{#lambda.pascalcase}}{{client}}{{/lambda.pascalcase}}Config(BaseConfig):
             {{/uniqueHost}}
             {{#hostWithAppID}}
             [
-                Host("{}-dsn.algolia.net".format(self.app_id), 10, CallType.READ),
-                Host("{}.algolia.net".format(self.app_id), 10, CallType.WRITE),
+                Host("{}-dsn.algolia.net".format(self.app_id), 'https', 10, CallType.READ),
+                Host("{}.algolia.net".format(self.app_id), 'https', 10, CallType.WRITE),
                 Host("{}-1.algolianet.com".format(self.app_id)),
                 Host("{}-2.algolianet.com".format(self.app_id)),
                 Host("{}-3.algolianet.com".format(self.app_id)),
```
Comment (shortcuts) [inline templates/python/config.mustache:58]: can you use named parameters and let the defaults do their thing? sorry it's a bit unrelated but seems less fragile

**Review by shortcuts — COMMENTED** [review]
looks really nice

File: clients/algoliasearch-client-php/lib/RetryStrategy/ApiWrapper.php (hunk 4 above)
Comment (damcou) [reply clients/algoliasearch-client-php/lib/RetryStrategy/ApiWrapper.php:157]: Yeah, if they want to specify a different scheme/port, they need to use `setFullHosts()` , I guess this would require some explanation indeed ...

File: templates/php/Configuration.mustache [hunk 7]
```diff
@@ -210,6 +211,18 @@ abstract class Configuration
...
+    {
+        $this->config['hasFullHosts'] = true;
+
+        return $this->setHosts($hosts);
+    }
+
+    public function getHasFullHosts()
+    {
+        return $this->config['hasFullHosts'];
+    }
```
[the remark this answers is not in the document]
Comment (millotp) [reply templates/php/Configuration.mustache:224]: I hate this but flemme to do more php

File: clients/algoliasearch-client-javascript/packages/client-common/src/types/host.ts [hunk 8]
```diff
@@ -13,6 +13,11 @@ export type Host = {
    * The protocol of the host URL.
    */
   protocol: 'http' | 'https';
```
[the remark this answers is not in the document]
Comment (shortcuts) [reply clients/algoliasearch-client-javascript/packages/client-common/src/types/host.ts:15]: idk the real definition of both I guess there's some overlap? to me scheme is `http` or `https` and protocol is how we transfer data, which is defined by the scheme?

idk idk up to you

File: templates/php/Configuration.mustache (hunk 7 above)
[the remark this answers is not in the document]
Comment (millotp) [reply templates/php/Configuration.mustache:224]: it can be improved in the future

File: clients/algoliasearch-client-javascript/packages/client-common/src/types/host.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (shortcuts) [reply clients/algoliasearch-client-javascript/packages/client-common/src/types/host.ts:15]: I prefer scheme anyway

[the remark this answers is not in the document]
Comment (millotp) [reply clients/algoliasearch-client-javascript/packages/client-common/src/types/host.ts:15]: to many idk ahah

[the remark this answers is not in the document]
Comment (shortcuts) [reply clients/algoliasearch-client-javascript/packages/client-common/src/types/host.ts:15]: ahahah that's my state of mind

File: templates/php/Configuration.mustache (hunk 7 above)
[the remark this answers is not in the document]
Comment (shortcuts) [reply templates/php/Configuration.mustache:224]: @damcou wdyt

[the remark this answers is not in the document]
Comment (damcou) [reply templates/php/Configuration.mustache:224]: I know this is not perfect like that but unfortunately we tried to use the same methotology as in the other languages, and it's not as simple since we reuse the Retry Strategy of the current PHP client and it would require quite a lot of rework :(

**Comment by shortcuts** [comment]
(btw gg to everyone who contributed to this, it's huge!)

File: templates/php/Configuration.mustache (hunk 7 above)
[the remark this answers is not in the document]
Comment (damcou) [reply templates/php/Configuration.mustache:224]: Let's merge this, this method will be only use for this test anyway.
